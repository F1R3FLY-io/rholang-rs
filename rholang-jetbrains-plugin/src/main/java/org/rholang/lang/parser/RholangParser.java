package org.rholang.lang.parser;

import com.intellij.lang.ASTNode;
import com.intellij.lang.PsiBuilder;
import com.intellij.lang.PsiParser;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.psi.tree.IElementType;
import org.jetbrains.annotations.NotNull;

public class RholangParser implements PsiParser {
    private static final Logger LOG = Logger.getInstance(RholangParser.class);

    @NotNull
    @Override
    public ASTNode parse(@NotNull IElementType root, @NotNull PsiBuilder builder) {
        PsiBuilder.Marker rootMarker = builder.mark();

        // Get the full text from the builder
        String text = builder.getOriginalText().toString();

        // Check if the code is valid using the Rust parser
        boolean isValid = RholangParserCli.isValid(text);
        LOG.debug("Rholang code is valid: " + isValid);

        // For now, just consume all tokens
        // In a more complete implementation, we would parse the tree returned by the Rust parser
        // and build a corresponding PSI tree
        while (!builder.eof()) {
            builder.advanceLexer();
        }

        rootMarker.done(root);
        return builder.getTreeBuilt();
    }
}
