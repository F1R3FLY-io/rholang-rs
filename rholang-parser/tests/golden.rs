use rholang_parser::RholangParser;
use rstest::rstest;
use std::fs;
use std::path::PathBuf;

#[rstest]
fn golden_test(
    #[base_dir = "tests/corpus/"]
    #[files("*.rho")]
    path: PathBuf,
) {
    let mut settings = insta::Settings::new();
    settings.set_snapshot_path("corpus/golden_snapshots");

    settings.bind(|| {
        let name = path.file_stem().unwrap().to_string_lossy().to_string();

        let parser = RholangParser::new();
        let input = fs::read_to_string(path).expect("Failed to read input file");
        let result = parser.parse(input.as_str());

        // Each snapshot is named after the input file
        insta::assert_debug_snapshot!(name, result);
    })
}
