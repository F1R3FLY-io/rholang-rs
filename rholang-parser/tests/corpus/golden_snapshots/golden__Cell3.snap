---
source: rholang-parser/tests/golden.rs
expression: result
---
Fail(
    [
        ParsingFailure {
            partial_tree: Some(
                AnnProc {
                    proc: Contract {
                        name: AnnName {
                            name: ProcVar(
                                Id(
                                    Id {
                                        name: "Cell",
                                        pos: SourcePos {
                                            line: 10,
                                            col: 10,
                                        },
                                    },
                                ),
                            ),
                            span: SourceSpan {
                                start: SourcePos {
                                    line: 10,
                                    col: 10,
                                },
                                end: SourcePos {
                                    line: 10,
                                    col: 14,
                                },
                            },
                        },
                        formals: Names {
                            names: [
                                AnnName {
                                    name: ProcVar(
                                        Id(
                                            Id {
                                                name: "client",
                                                pos: SourcePos {
                                                    line: 10,
                                                    col: 16,
                                                },
                                            },
                                        ),
                                    ),
                                    span: SourceSpan {
                                        start: SourcePos {
                                            line: 10,
                                            col: 16,
                                        },
                                        end: SourcePos {
                                            line: 10,
                                            col: 22,
                                        },
                                    },
                                },
                                AnnName {
                                    name: ProcVar(
                                        Id(
                                            Id {
                                                name: "state",
                                                pos: SourcePos {
                                                    line: 10,
                                                    col: 24,
                                                },
                                            },
                                        ),
                                    ),
                                    span: SourceSpan {
                                        start: SourcePos {
                                            line: 10,
                                            col: 24,
                                        },
                                        end: SourcePos {
                                            line: 10,
                                            col: 29,
                                        },
                                    },
                                },
                            ],
                            remainder: None,
                        },
                        body: AnnProc {
                            proc: ForComprehension {
                                receipts: [
                                    [
                                        Linear {
                                            lhs: Names {
                                                names: [
                                                    AnnName {
                                                        name: ProcVar(
                                                            Id(
                                                                Id {
                                                                    name: "request",
                                                                    pos: SourcePos {
                                                                        line: 11,
                                                                        col: 8,
                                                                    },
                                                                },
                                                            ),
                                                        ),
                                                        span: SourceSpan {
                                                            start: SourcePos {
                                                                line: 11,
                                                                col: 8,
                                                            },
                                                            end: SourcePos {
                                                                line: 11,
                                                                col: 15,
                                                            },
                                                        },
                                                    },
                                                ],
                                                remainder: None,
                                            },
                                            rhs: Simple {
                                                name: AnnName {
                                                    name: ProcVar(
                                                        Id(
                                                            Id {
                                                                name: "client",
                                                                pos: SourcePos {
                                                                    line: 11,
                                                                    col: 19,
                                                                },
                                                            },
                                                        ),
                                                    ),
                                                    span: SourceSpan {
                                                        start: SourcePos {
                                                            line: 11,
                                                            col: 19,
                                                        },
                                                        end: SourcePos {
                                                            line: 11,
                                                            col: 25,
                                                        },
                                                    },
                                                },
                                            },
                                        },
                                    ],
                                    [
                                        Linear {
                                            lhs: Names {
                                                names: [
                                                    AnnName {
                                                        name: ProcVar(
                                                            Id(
                                                                Id {
                                                                    name: "v",
                                                                    pos: SourcePos {
                                                                        line: 11,
                                                                        col: 27,
                                                                    },
                                                                },
                                                            ),
                                                        ),
                                                        span: SourceSpan {
                                                            start: SourcePos {
                                                                line: 11,
                                                                col: 27,
                                                            },
                                                            end: SourcePos {
                                                                line: 11,
                                                                col: 28,
                                                            },
                                                        },
                                                    },
                                                ],
                                                remainder: None,
                                            },
                                            rhs: Simple {
                                                name: AnnName {
                                                    name: ProcVar(
                                                        Id(
                                                            Id {
                                                                name: "state",
                                                                pos: SourcePos {
                                                                    line: 11,
                                                                    col: 32,
                                                                },
                                                            },
                                                        ),
                                                    ),
                                                    span: SourceSpan {
                                                        start: SourcePos {
                                                            line: 11,
                                                            col: 32,
                                                        },
                                                        end: SourcePos {
                                                            line: 11,
                                                            col: 37,
                                                        },
                                                    },
                                                },
                                            },
                                        },
                                    ],
                                ],
                                proc: AnnProc {
                                    proc: Match {
                                        expression: AnnProc {
                                            proc: ProcVar(
                                                Id(
                                                    Id {
                                                        name: "get",
                                                        pos: SourcePos {
                                                            line: 13,
                                                            col: 7,
                                                        },
                                                    },
                                                ),
                                            ),
                                            span: SourceSpan {
                                                start: SourcePos {
                                                    line: 13,
                                                    col: 7,
                                                },
                                                end: SourcePos {
                                                    line: 13,
                                                    col: 10,
                                                },
                                            },
                                        },
                                        cases: [
                                            Case {
                                                pattern: AnnProc {
                                                    proc: ProcVar(
                                                        Id(
                                                            Id {
                                                                name: "rtn",
                                                                pos: SourcePos {
                                                                    line: 13,
                                                                    col: 12,
                                                                },
                                                            },
                                                        ),
                                                    ),
                                                    span: SourceSpan {
                                                        start: SourcePos {
                                                            line: 13,
                                                            col: 12,
                                                        },
                                                        end: SourcePos {
                                                            line: 13,
                                                            col: 15,
                                                        },
                                                    },
                                                },
                                                proc: AnnProc {
                                                    proc: Bad,
                                                    span: SourceSpan {
                                                        start: SourcePos {
                                                            line: 14,
                                                            col: 9,
                                                        },
                                                        end: SourcePos {
                                                            line: 14,
                                                            col: 39,
                                                        },
                                                    },
                                                },
                                            },
                                            Case {
                                                pattern: AnnProc {
                                                    proc: ProcVar(
                                                        Id(
                                                            Id {
                                                                name: "newValue",
                                                                pos: SourcePos {
                                                                    line: 16,
                                                                    col: 12,
                                                                },
                                                            },
                                                        ),
                                                    ),
                                                    span: SourceSpan {
                                                        start: SourcePos {
                                                            line: 16,
                                                            col: 12,
                                                        },
                                                        end: SourcePos {
                                                            line: 16,
                                                            col: 20,
                                                        },
                                                    },
                                                },
                                                proc: AnnProc {
                                                    proc: Bad,
                                                    span: SourceSpan {
                                                        start: SourcePos {
                                                            line: 17,
                                                            col: 9,
                                                        },
                                                        end: SourcePos {
                                                            line: 17,
                                                            col: 34,
                                                        },
                                                    },
                                                },
                                            },
                                        ],
                                    },
                                    span: SourceSpan {
                                        start: SourcePos {
                                            line: 12,
                                            col: 4,
                                        },
                                        end: SourcePos {
                                            line: 19,
                                            col: 4,
                                        },
                                    },
                                },
                            },
                            span: SourceSpan {
                                start: SourcePos {
                                    line: 11,
                                    col: 3,
                                },
                                end: SourcePos {
                                    line: 20,
                                    col: 2,
                                },
                            },
                        },
                    },
                    span: SourceSpan {
                        start: SourcePos {
                            line: 10,
                            col: 1,
                        },
                        end: SourcePos {
                            line: 20,
                            col: 2,
                        },
                    },
                },
            ),
            errors: [
                AnnParsingError {
                    error: SyntaxError {
                        sexp: "(ERROR (var) (var))",
                    },
                    span: SourceSpan {
                        start: SourcePos {
                            line: 12,
                            col: 10,
                        },
                        end: SourcePos {
                            line: 12,
                            col: 22,
                        },
                    },
                },
                AnnParsingError {
                    error: SyntaxError {
                        sexp: "(ERROR)",
                    },
                    span: SourceSpan {
                        start: SourcePos {
                            line: 13,
                            col: 16,
                        },
                        end: SourcePos {
                            line: 13,
                            col: 17,
                        },
                    },
                },
                AnnParsingError {
                    error: SyntaxError {
                        sexp: "(ERROR (par (par (send channel: (var) send_type: (send_single) inputs: (inputs (var))) (send channel: (var) send_type: (send_single) inputs: (inputs (var)))) (var)))",
                    },
                    span: SourceSpan {
                        start: SourcePos {
                            line: 14,
                            col: 9,
                        },
                        end: SourcePos {
                            line: 14,
                            col: 39,
                        },
                    },
                },
                AnnParsingError {
                    error: SyntaxError {
                        sexp: "(ERROR (var))",
                    },
                    span: SourceSpan {
                        start: SourcePos {
                            line: 16,
                            col: 7,
                        },
                        end: SourcePos {
                            line: 16,
                            col: 10,
                        },
                    },
                },
                AnnParsingError {
                    error: SyntaxError {
                        sexp: "(ERROR (par (send channel: (var) send_type: (send_single) inputs: (inputs (var))) (var)))",
                    },
                    span: SourceSpan {
                        start: SourcePos {
                            line: 17,
                            col: 9,
                        },
                        end: SourcePos {
                            line: 17,
                            col: 34,
                        },
                    },
                },
                AnnParsingError {
                    error: MissingToken(
                        "}",
                    ),
                    span: SourceSpan {
                        start: SourcePos {
                            line: 20,
                            col: 2,
                        },
                        end: SourcePos {
                            line: 20,
                            col: 2,
                        },
                    },
                },
            ],
        },
    ],
)
